File: app/src/main/java/com/loopeer/codereader/ui/activity/RepositoryFragment.java
package com.loopeer.codereader.ui.activity;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.loopeer.codereader.R;
import com.loopeer.codereader.api.ServiceFactory;
import com.loopeer.codereader.api.service.GithubService;
import com.loopeer.codereader.model.Repository;
import com.loopeer.codereader.ui.adapter.RepositoryAdapter;
import com.loopeer.codereader.ui.decoration.DividerItemDecoration;
import com.loopeer.codereader.ui.fragment.BaseFragment;
import com.loopeer.codereader.utils.PageLinkParser;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class RepositoryFragment extends BaseFragment {

    private static final int PAGE_SIZE = 10;

    @BindView(R.id.view_recycler)
    RecyclerView mViewRecycler;

    private RepositoryAdapter mRepositoryAdapter;
    private GithubService mGithubService;

    private String mSearchText;

    private List<Repository> mRepositories = new ArrayList<>();

    private PageLinkParser mPageLinkParser;

    private boolean mIsLoading;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mRepositoryAdapter = new RepositoryAdapter(getActivity());
        mGithubService = ServiceFactory.getGithubService();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.activity_search_result, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupRecyclerView();
    }

    private void setupRecyclerView() {
        mViewRecycler.setLayoutManager(new LinearLayoutManager(getActivity()));
        mViewRecycler.setAdapter(mRepositoryAdapter);
        mViewRecycler.addItemDecoration(new DividerItemDecoration(getContext()));

        mViewRecycler.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
                if (layoutManager.findLastVisibleItemPosition() == mRepositoryAdapter.getItemCount() - 1 && isHasMore()) {
                    if (!mIsLoading)
                        requestData(mPageLinkParser.getNext());
                }
            }
        });
    }

    public void setSearchText(String searchText) {
        mSearchText = searchText;
        requestData(1);
    }

    public boolean isHasMore() {
        if (mPageLinkParser != null && mPageLinkParser.getNext() != 0 && mPageLinkParser.getRemain() != 0)
            return true;
        return false;
    }

    private void requestData(int page) {
        mIsLoading = true;

        if (page == 1) {
            mRepositories.clear();
            showProgressLoading("");
        }

        registerSubscription(mGithubService.repositories(mSearchText, null, null, page, PAGE_SIZE)
                .filter(baseListResponseResponse -> baseListResponseResponse.isSuccessful())
                .map(baseListResponseResponse -> {
                    mPageLinkParser = new PageLinkParser(baseListResponseResponse);
                    return baseListResponseResponse.body().items;
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(repositories -> {
                    mRepositories.addAll(repositories);
                    mRepositoryAdapter.setHasMore(isHasMore());
                    mRepositoryAdapter.updateData(mRepositories);
                    dismissProgressLoading();

                    mIsLoading = false;
                }, throwable -> {
                    throwable.printStackTrace();
                    dismissProgressLoading();

                    mIsLoading = false;
                }));
    }

    @Override
    public void showProgressLoading(String message) {
        mViewRecycler.setVisibility(View.INVISIBLE);
        super.showProgressLoading(message);
        getProgressLoading().setCanceledOnTouchOutside(false);
    }

    @Override
    public void dismissProgressLoading() {
        mViewRecycler.setVisibility(View.VISIBLE);
        super.dismissProgressLoading();
    }

}

File: code-reader-kt/src/main/java/com/loopeer/codereaderkt/ui/activity/RepositoryFragment.kt
package com.loopeer.codereaderkt.ui.activity

import android.os.Build
import android.os.Bundle
import android.support.annotation.Nullable
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.loopeer.codereaderkt.R
import com.loopeer.codereaderkt.api.ServiceFactory
import com.loopeer.codereaderkt.api.service.GithubService
import com.loopeer.codereaderkt.model.Repository
import com.loopeer.codereaderkt.ui.adapter.RepositoryAdapter
import com.loopeer.codereaderkt.ui.decoration.DividerItemDecoration
import com.loopeer.codereaderkt.ui.fragment.BaseFragment
import com.loopeer.codereaderkt.utils.PageLinkParser
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers


class RepositoryFragment : BaseFragment() {
    //TODO打开时会崩溃，未找到原因； 原因是BaseFragment继承的Fragment应该引入v4.Fragment而不是app.Fragment

    private val PAGE_SIZE = 10
    private lateinit var mViewRecycler: RecyclerView

    private lateinit var mRepositoryAdapter: RepositoryAdapter
    private lateinit var mGithubService: GithubService

    private lateinit var mSearchText: String

    private lateinit var mRepositories: ArrayList<Repository>

    private lateinit var mPageLinkParser: PageLinkParser

    private var mIsLoading: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mRepositories = ArrayList()
        mRepositoryAdapter = RepositoryAdapter(activity)
        mGithubService = ServiceFactory().getGithubService()
    }

    override fun onCreateView(inflater: LayoutInflater?, @Nullable container: ViewGroup?, @Nullable savedInstanceState: Bundle?): View? =
            inflater?.inflate(R.layout.activity_search_result, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mViewRecycler = view.findViewById<RecyclerView>(R.id.view_recycler)
        setupRecyclerView()
    }

    private fun setupRecyclerView() {
        mViewRecycler.layoutManager = LinearLayoutManager(activity)
        mViewRecycler.adapter = mRepositoryAdapter
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            mViewRecycler.addItemDecoration(DividerItemDecoration(context))
        }

        mViewRecycler.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                val layoutManager = recyclerView.layoutManager as LinearLayoutManager
                if (layoutManager.findLastVisibleItemPosition() == mRepositoryAdapter.itemCount - 1 && isHasMore()) {
                    if (!mIsLoading)
                        requestData(mPageLinkParser.next)
                }
            }
        })
    }

    fun setSearchText(searchText: String) {
        mSearchText = searchText
        requestData(1)
    }

    fun isHasMore(): Boolean = mPageLinkParser.next != 0 && mPageLinkParser.remain != 0

    private fun requestData(page: Int) {
        mIsLoading = true

        if (page == 1) {
            mRepositories.clear()
            showProgressLoading("")
        }

        registerSubscription(mGithubService.repositories(mSearchText, null.toString(), null.toString(), page, PAGE_SIZE)
                .filter { baseListResponseResponse -> baseListResponseResponse.isSuccessful }
                .map<List<Repository>> { baseListResponseResponse ->
                    mPageLinkParser = PageLinkParser(baseListResponseResponse)
                    baseListResponseResponse.body().items
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ repositories ->
                    mRepositories.addAll(repositories)
                    mRepositoryAdapter.setHasMore(isHasMore())
                    mRepositoryAdapter.updateData(mRepositories)
                    dismissProgressLoading()

                    mIsLoading = false
                }) { throwable ->
                    throwable.printStackTrace()
                    dismissProgressLoading()

                    mIsLoading = false
                })
    }

    override fun showProgressLoading(message: String) {
        mViewRecycler.visibility = View.INVISIBLE
        super.showProgressLoading(message)
        getProgressLoaing()?.setCanceledOnTouchOutside(false)
    }

    override fun dismissProgressLoading() {
        mViewRecycler.visibility = View.VISIBLE
        super.dismissProgressLoading()
    }


}